### PR TITLE
Rework Channel & Client closing signals

### PR DESCRIPTION
**Pull request by masad-frost**
Basically the idea is that the current channel and client closing handling sucks. More detail here https://github.com/replit/crosis/issues/18

Client:
`client.on('close'` now emits a `CloseResult` with a close reason. Diff: https://github.com/replit/crosis/commit/367ed3cbd67d08e4a7176acd5ea85fd76b7de291

Channel Close:
`channel.on('close'` now emits a `ChannelCloseReason`
Similarly `channel.close()` now resolves with `ChannelCloseReason`

Channel Request:
And the biggest change and the real impetus for this PR is that `channel.request` now doesn't throw an error when the channel closes during a request. Instead instead of resolving an `api.Command` we resolve an extended format of `api.Command` that has an optional `channelClosed` property. See `type RequestResult`

Test plan
=========
**Client:**
1. Connect client
2. attach `close` listener to client
3. call `client.close`
4. listener gets the expected close result

Repeat above, but step 3: Close websocket abruptly

**Channel Close:**
1. Connect client
2. Open channel
3. Wait for channel to open
4. Attach `close` listener on channel
4. Do steps for client closing above
5. See expected result

Repeat above, but step 4: call `channel.close`

Channel Request:
Basically do the same as above, but before closing the channel we call `channel.request`. We expect the request result to have a truthy `channelClosed` result and containers the right reasons.

### CODE REVIEW

**Comment by masad-frost** [comment]
Gonna merge and ship this under v5.0.0
